src/spi.rs

```diff
@@ -205,7 +224,8 @@ pub mod config {
 pub struct SpiBusMasterDriver<'d> {
     handle: spi_device_handle_t,
     trans_len: usize,
-    _p: PhantomData<&'d mut ()>,
+    hardware_cs: bool,
```

Comment:
Is this actually being used?

Reply:
the hardware_cs flag? if the user dont give an cs pin we need to unset the TRANS_CS_KEEP_ACTIVE otherwise we panic if we have cspin: -1

Reply:
this keeps track of it

Reply:
Yes.

Reply:
Ah I missed this in the diff. Would be nice if it could be a `const` generic to save on doing the checks at runtime.

Reply:
The generics in this case would largely be internal. And on top of saving one byte we'd be skipping a lot of checks to see whether `SPI_KEEP_CS_ACTIVE` needs to be used. This is especially true when Dma is not enabled. 

It's not possible anyway since cs/no-cs is specified at runtime. It would need to be different types for no-cs and cs devices.

Reply:
> The generics in this case would largely be internal. And on top of saving one byte we'd be skipping a lot of checks to see whether `SPI_KEEP_CS_ACTIVE` needs to be used. This is especially true when Dma is not enabled.
> 
> It's not possible anyway since cs/no-cs is specified at runtime. It would need to be different types for no-cs and cs devices.

It is possible, especially in light of my suggestion to have two separate public methods - for the cs/no-cs case each - but I'm still not convinced that checking a bool at runtime is a high cost to pay. What about the tons of other code, including  calls of nested functions, which will not be inlined with optimization = "s" and especially with the Rust-to-C FFI boundary that is always there? Are we mopping the grass with a nail cutter here? :p

Reply:
Oh - and how come the generics will be internal? Your type signature moves from `SpiDeviceDriver<'d>` into `SpiDeviceDriver<'d, const HW_CS_PIN: bool>`. Completely visible to the user, if I'm not missing anything.

Reply:
Ah, I meant `SpiDeviceDriver<'d>` would become `SpiDeviceDriver<'d>` and `SpiDeviceDriverNoCs<'d>`, instead of adding a generic parameter.
`SpiBusMasterDriver<'d>` would have an extra parameter but it would be an associated type of `SpiDeviceDriver<'d>` and `SpiDeviceDriverNoCs<'d>`, which is why it is mostly internal. User would not see it (unless they use a nice IDE).

I'm not gonna push for this since it's a micro optimisation that doesn't matter in the big picture, as you say. This is me officially dropping it.